sdk/storage/azure-storage-queue/src/main/java/com/azure/storage/queue/QueueAsyncClient.java

```diff
@@ -776,16 +938,55 @@ public HttpPipeline getHttpPipeline() {
         Function<String, Mono<PagedResponse<PeekedMessageItem>>> retriever =
             marker -> StorageImplUtils.applyOptionalTimeout(this.client.getMessages()
                 .peekWithResponseAsync(queueName, maxMessages, null, null, context), timeout)
-                .map(response -> new PagedResponseBase<>(response.getRequest(),
-                    response.getStatusCode(),
-                    response.getHeaders(),
-                    response.getValue(),
-                    null,
-                    response.getDeserializedHeaders()));
+                .flatMap(this::transformMessagesPeekResponse);
 
         return new PagedFlux<>(() -> retriever.apply(null), retriever);
     }
 
+    private Mono<PagedResponseBase<MessagesPeekHeaders, PeekedMessageItem>> transformMessagesPeekResponse(
+        MessagesPeekResponse response) {
+        List<PeekedMessageItemInternal> peekedMessageInternalItems = response.getValue();
+        List<PeekedMessageItem> peekedMessageItems = Collections.emptyList();
+        Mono<Void> mono = Mono.empty();
+        if (peekedMessageInternalItems != null) {
+            peekedMessageItems = new ArrayList<>(peekedMessageInternalItems.size());
+            for (PeekedMessageItemInternal peekedMessageItemInternal : peekedMessageInternalItems) {
+                try {
+                    peekedMessageItems.add(transformPeekedMessageItemInternal(
+                        peekedMessageItemInternal, messageEncoding));
+                } catch (IllegalArgumentException e) {
+                    if (messageDecodingFailedHandler != null) {
+                        mono = mono.then(messageDecodingFailedHandler.apply(
+                            new QueueMessageDecodingFailure(
+                                this,
+                                null,
+                                transformPeekedMessageItemInternal(
+                                    peekedMessageItemInternal, QueueMessageEncoding.NONE))
+                        ));
+                    } else {
```

Comment:
When calling the user's error handler, we should include the exception information that was thrown when decoding.